compiler/src/dotty/tools/dotc/core/StagingContext.scala

```diff
@@ -18,8 +24,27 @@ object StagingContext {
   def quoteContext(implicit ctx: Context): Context =
     ctx.fresh.setProperty(QuotationLevel, level + 1)
 
+  /** Context with an incremented quotation level and pushes a refecence to a QuoteContext on the quote context stack */
+  def pushQuoteContext(qctxRef: tpd.Tree)(implicit ctx: Context): Context =
+    val old = ctx.property(QuoteQontextStack).getOrElse(List.empty)
+    ctx.fresh.setProperty(QuotationLevel, level + 1)
+             .setProperty(QuoteQontextStack, qctxRef :: old)
```

Comment:
I'm wondering if the stack can be avoided using scoping rules for implicits.

Reply:
Yes, it could but we would need to track a level context and a splice context. The errors when those implicit are not found would also be hard to interpret.